### ISSUE TITLE
About Discriminator

### ISSUE THREAD

**Issue by ryanbbd** [christopher-beckham/hologan-pytorch#7]
Hi, glad to see the pytorch version! Looks like you are not using the WGAN or gp or others, just the adversarial loss. 
I run your code on car dataset, just  add the WGAN-GP term, things looks not so nice. 
Does the official tensorflow version use WGAN or other terms? Maybe I should remove this term and try again.

**Comment by christopher-beckham** [comment]
Hi,

The adversarial loss is JSGAN but it's with a spectrally normalised (SN) discriminator, as described [here](https://arxiv.org/abs/1802.05957). It is relatively SOTA-ish, and should perform better than vanilla WGAN/WGAN-SN/WGAN-GP (though it seems like hinge loss performs slightly better, so I should add that as an option). 

I have tried out an experiment with some publicly available cars dataset (see my tweet [here](https://twitter.com/chris_j_beckham/status/1259619607838699521)), but I will need to update the repo to add it as an option. Here were the sets of hyperparameters I tried for that particular experiment:

```
"angles": [ "[0,0,-90,90,0,0]" ],
"ngf": [512],
"ndf": [64],
"batch_size": [32],
"z_dim": [128],
"beta1": [0.5],
"beta2": [0.999],
"lamb": [1.0, 10.0, 25.0, 50.0],
"update_g_every": [1]
```

I would leave the code unchanged and simply try out these hyperparameters, but on your specific cars dataset.

Let me know if that helps

**Comment by ryanbbd** [comment]
Looks nice in you tweet! How long did this car experiment spend?
In my implementation, the g_loss looks a little bigger than d_loss, maybe it lose the adversarial effect.
I will try your sets of hyperparameters. Anyway, thanks for your feedback!

**Comment by christopher-beckham** [comment]
Hi,

Here's what I dug up (I think the individual curves here correspond to the four lambdas I posted above):

![image](https://user-images.githubusercontent.com/2417792/86271875-acb3a200-bb9b-11ea-8e33-7a7b0b212c63.png)

It is normal for the generator loss to be much higher. For this specific formulation of GAN (assuming your generator and discriminators have enough modelling capacity) you want to see `train_d_loss << train_g_loss`. Usually you will find that the generator loss goes up very slowly (and likewise for discriminator, it goes down very slowly) but that is normal. You can see that effect here from ~200 epochs onwards.

The red and orange curves are the most promising (sorry, not sure what lambdas they correspond to) since they also have small values for `train_g_z_loss` and `train_g_t_loss` (the losses that try to predict the `z` code and `theta` so as to maximise mutual information between `G(z; theta)` and `(z, theta)`.

Each epoch took about 110 seconds (it was a very small dataset), so for 800 epochs this corresponded to ~24.4 hours of training.

**Comment by christopher-beckham** [comment]
Hi @ryanbbd ,

I made a commit adding support for the Stanford cars dataset. There is a corresponding section for it in the README.md and an example script in `exps` you can run.

**Comment by ryanbbd** [comment]
Hi, @christopher-beckham ,
Very thanks for giving me such great reply! I will try your sets now! Keep in touch！
